File: src/puzzledata.cpp
#include "puzzledata.h"

#include "color.h"

PuzzleData::PuzzleData(const string &name)
	:m_name(name),
	 m_permutationList(),
	 m_slotList(),
	 m_stickerList(),
	 m_permutationFactory(new PermutationFactory()),
	 m_slotFactory(),
	 m_puzzleRenderer(NULL),
	 m_arcMap(),
	 m_center(0.0, 0.0)
{
}

void PuzzleData::addPermutation(Permutation *permutation)
{
	m_permutationList.push_back(permutation);
	m_arcMap[permutation] = vector<Arc *>(getSlotListSize(), NULL);

	for (unsigned int i = 0; i < getSlotListSize(); i++)
	{
		Slot *slotU = m_slotList[i];
		Slot *slotV = m_slotList[permutation->next(i)];
		Arc *lineArc = new LineArc(permutation, slotU, slotV);
		setArc(permutation, i, lineArc);
	}
}

void PuzzleData::render() const
{
	if (m_puzzleRenderer == NULL) return;
	
	m_puzzleRenderer->render(*this);
}

void PuzzleData::makeTestLevel()
{
	Color red(      "red", 1.0, 0.0, 0.0);
	Color yellow("yellow", 1.0, 1.0, 0.0);
	Color green(  "green", 0.0, 1.0, 0.0);
	Color blue(    "blue", 0.0, 0.0, 1.0);

	Slot *slot0 = m_slotFactory.makeSlot(red,    vec2f( 0.0, -3.0));
	Slot *slot1 = m_slotFactory.makeSlot(yellow, vec2f( 0.0,  0.0));
	Slot *slot2 = m_slotFactory.makeSlot(green,  vec2f(-3.0,  3.0));
	Slot *slot3 = m_slotFactory.makeSlot(blue,   vec2f( 3.0,  3.0));

	Slot *slotArray[] = { slot0, slot1, slot2, slot3 };

	for (Slot *slot : slotArray)
	{
		addSlot(slot);
		Sticker *slotSticker = slot->getSticker();
		slotSticker->setSlot(slot);
		addSticker(slotSticker);
	}

	vector<vector<unsigned int>> cycles0({ { 0, 1 }, { 2 }, { 3 } });
	vector<vector<unsigned int>> cycles1({ { 0 }, { 1, 2, 3 } });

	Permutation *p0 = m_permutationFactory->makePermutation(4, red,  0, cycles0);
	Permutation *p1 = m_permutationFactory->makePermutation(4, blue, 1, cycles1);

	addPermutation(p0);
	addPermutation(p1);
}

Arc *PuzzleData::getArc(Permutation *permutation, unsigned int nodeIndex, bool inverted) const
{
	if (m_arcMap.count(permutation) == 0) return NULL;
	const vector<Arc *> &arcList = m_arcMap.at(permutation);
	
	const vector<unsigned int> &mapping = permutation->getMapping();
	unsigned int nextIndex = mapping[nodeIndex];
	if (inverted)
	{
		for (unsigned int i = 0; i < mapping.size(); ++i)
			if (mapping[i] == nodeIndex)
				nextIndex = i;
	}
	
	return arcList[nextIndex];
}
		
void PuzzleData::setArc(Permutation *permutation, unsigned int nodeIndex, Arc *arc)
{
	if (m_arcMap.count(permutation) == 0)
		m_arcMap[permutation] = vector<Arc *>(getSlotListSize(), NULL);
	
	m_arcMap[permutation][nodeIndex] = arc;
}

File: src/js/puzzledata.js
class PuzzleData
{
	constructor(name)
	{
		this.name = name;
		this.permutationList = [];
		this.slotList = [];
		this.stickerList = [];
		this.arcMap = [];
		this.center = new vec2f(0.0, 0.0);
		this.easierList = [];
		this.harderList = [];
		this.localMap = [];
	}

	solve()
	{
		for (var i = 0; i < this.stickerList.length; i++)
		{
			var sticker = this.stickerList[i];
			sticker.cloneSlot();
		}
	}

	randomize()
	{
		var numPermutations = this.getPermutationListSize();
		for (var i = 0; i < 1000; i++)
		{
			var randomIndex = randInt(1, numPermutations) - 1;
			var randomInverted = randInt(1, 2) == 1;
			this.activatePermutation(randomIndex, randomInverted);
		}
	}

	activatePermutation(index, inverted)
	{
		if (index >= this.permutationList.length) return;

		var permutation = this.permutationList[index];
		permutation.apply(this, inverted);
	}

	draw(context)
	{
		drawString(context, '#000000', this.name, 5, 20);
		
		for (var i = 0; i < this.permutationList.length; i++)
		{
			var permutation = this.permutationList[i];
			permutation.draw(context, this.getArcList(permutation));
		}

		for (var i = 0; i < this.slotList.length; i++)
			this.slotList[i].draw(context);

		for (var i = 0; i < this.stickerList.length; i++)
			this.stickerList[i].draw(context);
	}

	addEasier(puzzleName)
	{
		this.easierList.push(puzzleName);
	}

	addHarder(puzzleName)
	{
		this.harderList.push(puzzleName);
	}

	addLocalMap(mapStr)
	{
		this.localMap.push(mapStr);
	}

	getEasierList()
	{
		return this.easierList;
	}

	getHarderList()
	{
		return this.harderList;
	}

	getLocalMap()
	{
		return this.localMap;
	}
	
	addPermutation(permutation)
	{
		this.permutationList.push(permutation)
		this.arcMap[permutation.index] = [];

		for (var i = 0; i < this.slotList.length; i++)
		{
			var slotU = this.slotList[i];
			var slotV = this.slotList[permutation.next(i)];
			var lineArc = NewLineArc(permutation, slotU, slotV);
			this.setArc(permutation, i, lineArc);
		}
	}

	addSlot(slot)
	{
		this.slotList.push(slot);
	}

	addSticker(sticker)
	{
		this.stickerList.push(sticker);
	}

	clearSlotList()
	{
		this.slotList = [];
	}

	clearStickerList()
	{
		this.stickerList = [];
	}

	getPermutationList()
	{
		return this.permutationList;
	}

	getSlotList()
	{
		return this.slotList;
	}

	getStickerList()
	{
		return this.stickerList;
	}

	getPermutationListSize()
	{
		return this.permutationList.length;
	}

	getSlotListSize()
	{
		return this.slotList.length;
	}

	getStickerListSize()
	{
		return this.stickerList.length;
	}

	setPermutationList(permutationList)
	{
		this.permutationList = permutationList;
	}

	setSlotList(slotList)
	{
		this.slotList = slotList;
	}

	setStickerList(stickerList)
	{
		this.stickerList = stickerList;
	}

	getPermutation(index)
	{
		return this.permutationList[index];
	}

	getSlot(index)
	{
		return this.slotList[index];
	}

	getSticker(index)
	{
		return this.stickerList[index];
	}

	getArc(permutation, nodeIndex, inverted)
	{
		if (permutation.index >= this.arcMap.length)
			return null;
		
		var arcList = this.arcMap[permutation.index];
		var mapping = permutation.getMapping();
		var nextIndex = mapping[nodeIndex];
		if (inverted)
		{
			for (var i = 0; i < mapping.length; i++)
				if (mapping[i] == nodeIndex)
					nextIndex = i;
		}

		return arcList[nextIndex];
	}

	setArc(permutation, nodeIndex, arc)
	{
		if (permutation.index >= this.arcMap.length)
			this.arcMap[permutation.index] = zeroArray(this.getSlotListSize());

		this.arcMap[permutation.index][nodeIndex] = arc;
	}

	getArcList(permutation)
	{
		return this.arcMap[permutation.index];
	}

	getCenter()
	{
		return this.center;
	}

	setCenter(center)
	{
		this.center = center;p
	}
}
